### ISSUE TITLE
HTML whitelist as external list

### ISSUE THREAD

**Issue by BsNoSi** [datenstrom/yellow#515]
I would like to suggest to extract the whitelist from the core to an external list, because this makes maintenance easier. Most importantly, this allows site specific customizing that "survives" an update.

In regard to several issues of the near past I do not believe in frequent update for adding items to the whitelist that leads step ty step to the previous "nothing blocked" behaviour.

**Comment by BsNoSi** [comment]
Generally a blacklisting would be much easier to deal with. 

It would be a significantly shorter list that can be handled much faster: only the few *unwanted* things are filtered instead of *wanted* are allowed to pass thru.

**Comment by BsNoSi** [comment]
**As long as extensions are touched by withelisting** much extensions are blocked without a core update that *allows* the use of required functionality for the extension. 

The consequences are 

- no chance to fix broken extensions because of whitelisting (except much more forking which hopefully is not the target)
- much higher challenge to develop useful extensions
- much lower interest to write extensions for public
- (my personal feeling currently) annoyed developers/maintainer

**Comment by BsNoSi** [comment]
Maybe an external "additional to standard whitlist" would be a useful compromise.

The core holds the basic list (as given) and it is possible to add entries in a `whitelist.txt`:

```
[elementsHtml]
…
…
[elementsSvg]
…
…
[attributesHtml]
…
…
[attributesSvg]
…
…
```

This would assures a "safe" standard setup (very relevant for rookies) but a *core code save* solution for adding things at one's own risk.

Developers can use additional attributes they have to mention in their readme.md of extension.

As  [mentioned here](https://github.com/datenstrom/yellow/issues/517#issuecomment-629602185) besides this a settings parameter for disabling filtering could be useful, too.

Or to keep it save but communicative without need of debugging the switch leads to a note `»bla«  not allowed for security reasons, see …` in edit preview.

**Comment by GiovanniSalmeri** [comment]
TL;DR: I would have a slight preference for excluding the extensions from the whitelist filtering.

I have mixed feelings regarding the current implementation of whitelisting from the point of view of an author of some little extensions. On the one hand, it forces to write *theoretically* cleaner code (for example, <script> only in the head section, events programmatically assigned instead of onclick attribute ecc.): there is nothing sane that cannot be achieved with the current limitations (not all is sane: for example `document.write` is not...). On the other hand, in the recent weeks I stumbled at least four or five times across these limitations and 1. each time I wondered  whether the final result was indeed *cleaner* or rather employed workarounds which make the code less clear 2. the casual author of a site will find a gazillion places that instruct to put a <script> in the body (by modifying a template, for example): this will simply not work in Yellow (two examples from my recent experience: https://github.com/GiovanniSalmeri/yellow-replain, https://github.com/datenstrom/yellow/issues/499#issuecomment-602179761), and he will blame Yellow, not these instructions which work "evereywhere else."

Honestly I do not know which solution is better. But I would have a slight preference for at least excluding the extensions from the whitelist filtering: after all, extensions can be dangerous in many ways much, much worse than allowing a <script> tag in the body: so I see little advantage in imposing this limitation. (I have not studied the internals of core.php in order to know if this change in the current behaviour would be easy.) This obviously does not address the scenario of someone who wants to insert for example a script in the body section of a layout, but at least would avoid workarounds in the extensions.

**Comment by BsNoSi** [comment]
There are pros and cons to filtering of extensions.



| pro                                                          | con                                                          |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| Input of an extension by user is under control               | generated code of extension may be damaged                   |
| Possibly compromising code is removed                        | Possibly valuable/required code is removed                   |
| There is a "general security" that takes care of integrity of all generated code | Too high security settings restrict posibilities              |
| Security settings are centralised - all sites profit from the knowledge of a sopisticated developer | Security settings are centralised – all sites depend from the setup that is controlled by the core developer (fork may solve this but this splits development into pieces → less power for main fork) |

Additionally to my ideas above maybe a system function "addtowhitelist" for extension development may be a third way to allow things "not in the list" instead of a `whiteliste.txt`.

Or – forth and last idea – filtering is extracted into an own extension that is easier to modify and — if modified — does not block core from update or always has to be modified with individual changes after any update.

**Comment by markseu** [comment]
You can copy/paste stuff into the header/footer in your `system/layouts` folder, no questions asked.

For a smoother integration you have the possibility to make extensions, but this means you have to get familiar with our API and design patterns. In an ideal world the right things should be easy to do and the wrong things hard to do. It might be an unpopular opinion among web developers, but I like it when software prevent people from shooting into their own foot. The only purpose of the HTML filter is to normalise the data generated by the Markdown parser. If your extensions wants to inject extra data, for example into a page header/footer, no HTML filter is applied.

As maintainer I am going to say no to the idea of a configurable HTML whitelist. If you find out that something is missing in the HTML whitelist, please report it as a bug and we can fix it right away.

**Comment by BsNoSi** [comment]
First of all: Maybe some of my last entries sounded aggressive. Truly, I was annoyed by an effect that kept appearing. But this was my fault, not yours. I follow you explicitly with your attempt to make Yellow CMS "user save". Regardless to this I think it is one strength of open development that it is possible to think about the ways to achieve this.

As I wrote above as a pro »all sites profit from the knowledge of a sopisticated developer« – this is absolutely ok for me and truly a strong argument. Have you a short description which case of options or commands you expect to be harmful? 

Then we all can examine the current list in respect to your given "no's" and collect a list of possibly missing entries that might be added to the WhiteList. This could reduce your efforts tremendeously.

One is "onClick" for me because if you have css that stacks areas that shall link to a target is not possible with `<a …>…</a>`.  Currently I added it on my own to the core – but possibly you have a good reason to let it out – which I'd like to understand.

**Comment by markseu** [comment]
I don think there's a solution that makes everyone happy. I would separate HTML and JavaScript, for example remove `onclick` from HTML and handle `DOMContentLoaded` in JavaScript. Have a look at this [JavaScript tutorial about DOMContentLoaded](https://www.javascripttutorial.net/javascript-dom/javascript-domcontentloaded/) and [OWASP's XSS prevention cheat sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html).

**Comment by BsNoSi** [comment]
No one of us can honestly say »I have the best solution.« We have "points of view". I personally try to avoid JavaScript anyway. But sometimes it is needed to add functionality html/css does not offer.

I do need `onClick` to pass `"location.href='target'` because I did not find a way to achieve reliably access to links in stacked DIVs that are moved by css. The upmost `a` is not identified correctly in any browser, `onClick` is (no idea why, gave up after hours of trying around → long *before* filtering took place!). The use of `onClick` is for me a reluctant concession, because I want to have a benefit that can only be achieved with it in these circumstances.

In some cases `script` is required to have the possibility for creating dynamically script depending to the content of e.g. an image folder for action. This script is created by php on runtime and is not static. 

**Therefore, the question is, what Yellow CMS defines as »Allowed Locations«** (rule #0 cheat sheets). 

If you think that code generated by extensions that needed to be placed by an admin in their folder is generally harmful and therefore has to be filtered this is a very strong restriction for extending Yellow with useful/neat functions. Therfore, I am with @GiovanniSalmeri .

Concrete simple example:

What about `button` (part of $elementsHtml) ? 
If you want to use it "standalone" you need `onClick` for smart code:
```
<button onclick="window.location.href='https://google.com'">Go to Google</button>
```
 to avoid something like
```
<form action="https://google.com">
    <input type="submit" value="Go to Google" />
</form>
```
which bloats the code and – allowing `form` opens another door for problems (but also part  of $elementsHtml).

You may avoid `button` by styling `a`:
```
<sytle> 
 a.button {Appearance: button} 
</style>
…
<a href="https:/google.com" class="button">Go to Google</a>
```
but in this case you do not need a whitelisted `button`.

To make it clear again: I am absolutely with you to have a "secure user environment" in Yellow. The question is if the extensions folder is part of "user space" or is "sophisticated area".

To put it bluntly: If I have access to the extensions folder I can compromise the system anyway. Therefore the filtering of code generated with extensions in my point of view primarily hinders developers but does not have the targeted effect for the user space.

Because of this the question is where to draw the line. I honestly respect your attempts to hand out code that delivers secure results. Nevertheless I would like to appeal to you to rethink the filtering of extensions. Because by doing so, you are taking responsibility for the work of third parties, which I would not take over under any circumstances. Because it is solely and exclusively their responsibility what their extensions do.

Otherwise an impression of security could be created for which you objectively cannot guarantee in any way. The administrator of a website is always responsible for security.

**Comment by markseu** [comment]
Okay, thanks for all your feedback. As maintainer I am going to say no to the idea of a configurable HTML whitelist. This feature plays a central role in making secure websites. I think the way forward is to help each other. Let us know if you need help with the API. If you see a developer in trouble, please review the code and send a pull request.
